File: src/main/java/me/zeroeightsix/kami/module/modules/utils/EntityTools.java
package me.zeroeightsix.kami.module.modules.utils;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.RayTraceResult;
import net.minecraftforge.fml.common.gameevent.InputEvent;
import org.lwjgl.input.Mouse;

/**
 * @author 0x2E | PretendingToCode
 *
 * Inspired by ForgeHax, recreated with expressed permission from creator
 *
 * TODO: Fix delay timer because that shit broken
 */
@Module.Info(name = "EntityTools", category = Module.Category.UTILS, description = "Right click entities to perform actions on them")
public class EntityTools extends Module {
    private Setting<Mode> mode = register(Settings.e("Mode", Mode.DELETE));
    private int delay = 0;
    private enum Mode {
        DELETE, INFO
    }

    @Override
    public void onUpdate() {
        if (delay > 0) {
            delay--;
        }
    }

    @EventHandler
    public Listener<InputEvent.MouseInputEvent> mouseListener = new Listener<>(event -> {
        if (Mouse.getEventButton() == 1 && delay == 0) {
            if (mc.objectMouseOver.typeOfHit.equals(RayTraceResult.Type.ENTITY)) {
                if (mode.getValue().equals(Mode.DELETE)) {
                    mc.world.removeEntity(mc.objectMouseOver.entityHit);
                }

                if (mode.getValue().equals(Mode.INFO)) {
                    NBTTagCompound tag = new NBTTagCompound();
                    mc.objectMouseOver.entityHit.writeToNBT(tag);
                    Command.sendChatMessage(getChatName() + "&6Entity Tags:\n" + tag + "");
                }
            }
        }
    });
}

File: src/main/java/me/zeroeightsix/kami/module/modules/misc/EntityTools.kt
package me.zeroeightsix.kami.module.modules.misc

import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.ModuleConfig.setting
import me.zeroeightsix.kami.util.TickTimer
import me.zeroeightsix.kami.util.text.MessageSendHelper
import net.minecraft.entity.Entity
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.RayTraceResult
import net.minecraftforge.fml.common.gameevent.InputEvent
import org.kamiblue.event.listener.listener
import org.lwjgl.input.Mouse

object EntityTools : Module(
    name = "EntityTools",
    category = Category.MISC,
    description = "Right click entities to perform actions on them"
) {
    private val mode = setting("Mode", Mode.INFO)

    private enum class Mode {
        DELETE, INFO
    }

    private val timer = TickTimer()
    private var lastEntity: Entity? = null

    init {
        listener<InputEvent.MouseInputEvent> {
            if (Mouse.getEventButton() != 1 || mc.objectMouseOver == null || mc.objectMouseOver.typeOfHit != RayTraceResult.Type.ENTITY) return@listener
            if (timer.tick(5000L) || mc.objectMouseOver.entityHit != lastEntity && timer.tick(500L)) {
                when (mode.value) {
                    Mode.DELETE -> {
                        mc.world.removeEntity(mc.objectMouseOver.entityHit)
                    }
                    Mode.INFO -> {
                        val tag = NBTTagCompound().apply { mc.objectMouseOver.entityHit.writeToNBT(this) }
                        MessageSendHelper.sendChatMessage("""$chatName &6Entity Tags:$tag""".trimIndent())
                    }
                }
            }
        }
    }
}
